dipy/reconst/dti.py

```diff
@@ -1514,111 +1604,129 @@ def _ols_fit_matrix(design_matrix):
+            M-estimator is used for weighting, in which case this value
+            should be C (see below).
 
-    .. math ::
+        Notes
+        -----
+        The Geman-McClure M-estimator is described as follows [1]_ (page 1089):
+        "The scale factor C affects the shape of the GMM
+        [Geman-McClure M-estimator] weighting function and represents the expected
+        spread of the residuals (i.e., the SD of the residuals) due to Gaussian
+        distributed noise. The scale factor C can be estimated by many robust scale
+        estimators. We used the median absolute deviation (MAD) estimator because
+        it is very robust to outliers having a 50% breakdown point (6,7).
+        The explicit formula for C using the MAD estimator is:
 
-            C = 1.4826 x MAD = 1.4826 x median{|r1-\hat{r}|,... |r_n-\hat{r}|}
+        .. math ::
 
-    where $\hat{r} = median{r_1, r_2, ..., r_3}$ and n is the number of data
-    points. The multiplicative constant 1.4826 makes this an approximately
-    unbiased estimate of scale when the error model is Gaussian."
+                C = 1.4826 x MAD = 1.4826 x median{|r1-\hat{r}|,... |r_n-\hat{r}|}
 
+        where $\hat{r} = median{r_1, r_2, ..., r_3}$ and n is the number of data
+        points. The multiplicative constant 1.4826 makes this an approximately
+        unbiased estimate of scale when the error model is Gaussian."
 
-    References
-    ----------
-    .. [1] Chang, L-C, Jones, DK and Pierpaoli, C (2005). RESTORE: robust
-    estimation of tensors by outlier rejection. MRM, 53: 1088-95.
-    """
-    # This is the predicted signal given the params:
-    y = np.exp(np.dot(design_matrix, tensor))
-
-    # Compute the residuals
-    residuals = data - y
-
-    # If we don't want to weight the residuals, we are basically done:
-    if weighting is None:
-        # And we return the SSE:
-        return residuals
-    se = residuals ** 2
-    # If the user provided a sigma (e.g 1.5267 * std(background_noise), as
-    # suggested by Chang et al.) we will use it:
-    if weighting == 'sigma':
-        if sigma is None:
-            e_s = "Must provide sigma value as input to use this weighting"
-            e_s += " method"
-            raise ValueError(e_s)
-        w = 1 / (sigma**2)
 
-    elif weighting == 'gmm':
-        # We use the Geman-McClure M-estimator to compute the weights on the
-        # residuals:
-        C = 1.4826 * np.median(np.abs(residuals - np.median(residuals)))
+        References
+        ----------
+        .. [1] Chang, L-C, Jones, DK and Pierpaoli, C (2005). RESTORE: robust
+        estimation of tensors by outlier rejection. MRM, 53: 1088-95.
+        """
+        # This is the predicted signal given the params:
+        y = np.exp(np.dot(design_matrix, tensor))
+        self.y = y  # cache the results
+
+        # Compute the residuals
+        residuals = data - y
+        self.residuals = residuals  # cache the results
```

Comment:
why residuals in the cache ? the cached value seems to not be used

Reply:
sorry about that, left over from a previous setup where we were updating the GMM weights on each call of the error function (rather than updating them after each optimization, which is the correct thing to do)